exp/controllers/machinepool_controller_phases.go

```diff
@@ -108,6 +108,33 @@ func (r *MachinePoolReconciler) reconcileExternal(ctx context.Context, cluster *
 		return external.ReconcileOutput{Paused: true}, nil
 	}
 
+	if strings.HasSuffix(ref.Kind, external.TemplateSuffix) {
+		owner := &metav1.OwnerReference{
+			APIVersion: expv1.GroupVersion.String(),
+			Kind:       "MachinePool",
+			Name:       m.Name,
+			UID:        m.UID,
+		}
+		ref, err = external.CloneTemplate(ctx, &external.CloneTemplateInput{
+			Client:      r.Client,
+			TemplateRef: ref,
+			Namespace:   m.Namespace,
+			ClusterName: cluster.Name,
+			OwnerRef:    owner,
+		})
```

Comment:
/hold

Wouldn't this code clone the reference every time it reconciles?

Reply:
cc @CecileRobertMichon @fabriziopandini

Reply:
In my point of view no because reference is replaced by concrete object reference, so in next reconciliation the code will do not pass inside if block

Reply:
The `ref, err = external.CloneTemplate` assignment would only replace the reference for the current function, am I missing something?

Reply:
line 128 is getting the ref object and this object returned by function

Reply:
Yeah, but `obj` is a local variable, it doesn't get changed outside

Reply:
obj variable is returned to outside no? https://github.com/kubernetes-sigs/cluster-api/blob/1f87ca9cc54e5c22217bf09234a604450b1b91b1/exp/controllers/machinepool_controller_phases.go#L192

Reply:
Yeah, although, looking through the caller's code, I don't see that the reference on the MachinePool object ever gets updated after the first iteration, which is going to cause the templated object to be cloned again. We should probably have a test against this behavior as part of this PR.

Apart from the above, the behavior you were describing if I understood correctly is:

- MachinePool's references can be created with both template and non template objects
- If a template object is found, we want to clone it
- After cloning, we set the infrastructure or bootstrap reference to the new cloned object

The behavior describe above wouldn't be a common one in Cluster API, for example MachineDeployment and MachineSet use references to clone an object to create a different object. If we go down this path, we'd end up in a situation where the user submits an object with `infraRef` and `bootstrapRef` pointing to a template, but the controller overwrites those with a different reference. The first thing that comes to mind is that this will break gitops-like deployments.

One possible workaround is to add some additional fields that would reference the cloned object, or we could clone with pre-defined names and check if those objects exist before cloning again.

Reply:
@vincepri Good points. Looking at this side, I am almost convinced that supporting template and non-template objects is not a good idea mainly because we break gitops, but now I have a basic doubt. When to use template objects and When to use non-template objects?

Reply:
> When to use template objects and When to use non-template objects?

We could let Template objects to be used, but by convention, not configuration. What I mean by that is that we need a way to ensure that the Cloned object uses a predefined name, which can be based on the name of the MachinePool object. In the next reconciliation loop, we should make sure that if we see a template object, we try to retrieve the object with the predefined name first, and only if that object cannot be found, we go ahead and create it.

I'm not 100% sold on either solution, but it might be a start, @CecileRobertMichon @devigned what do you all think?

Reply:
The issue with relying on template name is that if we later decide to change that default name we wouldn't be able to find the objects that were created by a previous version of the controller...

I'm not 100% following what we're trying to achieve here. @felipeweb can you please add a PR description that explains your use case? The linked issue is in the CAPA provider and references a different issue as well. We should al[proposal](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20190919-machinepool-api.md) to reflect this change, as the current assumption in the proposal is that MachinePools reference 1 bootstrap config object since all the instances in the pool have the same user data.

@devigned will probably have more insight here as it sounds like it's something he's thought about recently, but he's out until Tuesday.

Reply:
To clarify my [earlier statement](https://github.com/kubernetes-sigs/cluster-api/pull/4112#pullrequestreview-579187865), the `MachinePools` and `MachineDeployments` incongruity with `KubeadmConfig` and `KubeadmConfigTemplate` respectively caused me some confusion when I first started writing cluster yamls for the `AzureMachinePool` implementation.

When I initially authored the cluster yamls for `AzureMachinePool`, I started with a working set of resources from the CAPZ default `cluster-template.yaml`. I started replacing the `MachineDeployment` with a `MachinePool` and kept the `KubeadmConfigTemplate`. After writing a bit of code, I started testing and found that CAPI `MachinePool` controller didn't pick up on the `KubeadmConfigTemplate`, but rather, was looking for a `KubeadmConfig`. It took a bit of debugging and getting into the internals of the `MachinePool` controller (*I was really new to the project at the time and still trying to wrap my head around the project.*)

However much confusion I had previously, I feel better about about the distinction today since, as @CecileRobertMichon pointed out, `MachinePools` reference 1 bootstrap config, not many instances of a config template. That distinction may not make sense to a newcomer to the project. They may try to write `MachinePools` in a similar manor as a `MachineDeployment` without regard to the subtle distinction between them.

I don't feel strongly one way or another. Just thought I'd share my early experiences in case it can provide some color.

Reply:
What @devigned described was exactly what I went through as well. I don't have a use case for this yet, only those pains that @devigned pointed out.

Reply:
Got it. In that case, how can we make this subtle distinction very explicit so it's not so confusing for newcomers? I don't think allowing templates and making MachinePool _more_ like MachineDeployments is the answer here, but we should make it clear why and _how_ they are different.

Reply:
Maybe pictures would tell the story our words have failed to tell.

Documentation showing a visual representation comparing a `MachineDeployment` and `MachinePool` would be super helpful. Perhaps, a multi-step visualization starting from initial apply and ending with deployment completed with multiple machines that describes the anatomy, relationships and choreography of the resources the user specified and the ones created by the controller. This documentation could be less developer focused and more user focused; less state diagrams and more context / intent.

@felipeweb do you think that would have helped you?

Reply:
@devigned sure!